proto/tpm_enrollz.proto

```diff
@@ -58,6 +58,11 @@ message RotateOIakCertRequest {
   // PEM-encoded owner initial DevID certificate signed by the
   // switch owner/administrator CA.
   string oidevid_cert = 3;
+
+  // SSL profile for which the certs will be assigned.
+  // (Optional: Only required if rotating oidevid_cert.)
+  string ssl_profile_id = 4;
```

Comment:
Note that this may not be necessary. 
Another way to model an iDevID/oiDevID-based profile is to introduce a flag/enum to the tls-profile configuration.

E.g. a profile can be modelled as
```
profile Bob {
certificate  
key  
trust_bundle
...
enum use_devid
}
```
and use_devid enum can have three values (`disable`, `oiDevid`, `iDevID`); or it can be expressed as bool `true`/`false` (if `true`, oiDevID will be used when available, iDevID otherwise)

In that case there's no need to specify the profile names explicitly, you can say that all profiles with the flag should use the updated cert data.

Reply:
>In that case there's no need to specify the profile names explicitly, you can say that all profiles with the flag should use the updated cert data.

I think we should be consistent - if we're using certz for SSL profile management, and certz operates on SSL profiles, then we should avoid creating some alternative representation of these profiles with extra attestz-specific flags. 

Unless you were considering adding this into the certz spec too, in which case im still not really sure its a good approach :)

Fundamentally the SSL profile is just a logical grouping of files (certs/keys/crls) which can be applied to a gRPC server. If we attach some additional flag to these files which specifies that "updates to this profile affect all other profiles" then we have this kind of transitive dependency which isnt immediately visible to the client performing a certz rotation. 
I.e. an update to SSL profile A now may have knock-on effects for some unknown SSL profile B and C

Reply:
might this conversation (all of it really) benefit from a video-chat to hash out the whys/where-fores?

Reply:
I would agree that there's no magic, but, at the same time, an iDevID or oiDevID profile is not the same as a profile fully configured via certZ, for instance.

Today, when a profile is defined via certZ, you supply all the info in the profile definition (certs, private key, trust-bundle, etc).  
For oiDevId, the key is not included in this equation since it is stored only in TPM.

Instead of making an implicit assumption that all profiles can be configured without keys, and the device should perform magic checks on each keyless profile to see if it could be an oiDevID profile, etc; our proposal is to make this part explicit.
Give a clear indication to the device "expect to use tpm with the profile". And that state can also be exposed via telemetry if needed.

Another thing is that we don't expect to have more than a single oiDevID cert on the device at the same time. And conceptually it is much closer to an iDevID cert (a single entity that is not stored as a part of any profile).

Reply:
good points @LimeHat 

> On the contrary, our proposal is to add an OpenConfig-compliant way to represent a tls-profile that uses iDevID or oiDevID.

Ah, I was focused on the narrow case where we just want to use attestz to configure this profile - in which case we can know "this rotation is occuring at a stage in the attestz workflow, so we should expect that they only transfer a cert and that we pull the key from the TPM".
I hadn't thought about the broader case where some vendor may (after bootz/attestz has completed) want to configure a profile like this.

>Instead of making an implicit assumption that all profiles can be configured without keys, and the device should perform magic checks on each keyless profile to see if it could be an oiDevID profile, etc; our proposal is to make this part explicit.

I do like the idea of making it explicit - this part sounds totally reasonable.

Perhaps would it make sense to add this `use_devid` enum into the `gnsi.certz.v1.Certificate` message type here https://github.com/openconfig/gnsi/blob/1996bef75e54512d49cb811a10e503db297aaba8/certz/certz.proto#L447-L461  as a more generic `KeySource` enum ?
For example, something like:
```proto 
message Certificate {
  // Type of certificate.
  CertificateType type = 1;
  // Certificate encoding type.
  CertificateEncoding encoding = 2;

  // Actual certificate.
  // The exact encoding depends upon the type of certificate.
  // for X509, this should be a PEM encoded Certificate.
  bytes certificate = 3;
  
  // Optional.
  // Not needed if the device knows the private key or does not have to know it.
  bytes private_key = 4;
  
  enum KeySource {
     KEYSOURCE_UNSPECIFIED = 0;
     KEYSOURCE_GENERATED = 1; // case where certz generated the key with a CSR
     KEYSOURCE_TRANSFER = 2; // case where key is transferred in this message ( private_key != nil )
     
     // keys corresponding to specific certs
     KEYSOURCE_OIDEVID = 3; 
     KEYSOURCE_IDEVID = 4; 
  }
  
  KeySource key_source = 5;
}
```
and then maybe we can re-use that certificate message here for the rotation?
i.e use:
```proto
message RotateOIakCertRequest {
  // Switch control card selected identifier.
  ControlCardSelection control_card_selection = 1;

  // PEM-encoded owner initial attestation key certificate signed by the
  // switch owner/administrator CA.
  gnsi.certz.v1.Certificate oiak_cert = 2;

  // PEM-encoded owner initial DevID certificate signed by the
  // switch owner/administrator CA.
  gnsi.certz.v1.Certificate oidevid_cert = 3;

  // SSL profile for which the certs will be assigned.
  // (Optional: Only required if rotating oidevid_cert.)
  string ssl_profile_id = 4;
}
```
That way we could still deal with the general case (beyond attestz) where we need to handle these types of certs in profiles

Reply:
can there be any other defined key sources than idevid or oidevid like are those truely the only "special cases"?

Reply:
also i would say in these cases for the private key we should probably make a proper oneof

oneof KeySource {
   bool csr_generated = 1;
   bytes private_key = 2;
   bool idevid = 3;
   bool oidevid = 4;
  }
  
  ?

Reply:
> can there be any other defined key sources than idevid or oidevid like are those truely the only "special cases"?

The IDevID key/cert are fixed known TPM location (handle) defined by each vendor, so saying idevid in config is enough for the switch to know what to do (different vendor will chose different handle but that's transparent).

TPM handles are memory location such as handle:0x81820000, this is how it would be used underneath in openssl commands for instance.

If a credential per card other than idevid was used, something user-provisioned, and if this was stored in the TPM then the way to address it would be by its actual handle location such as key = "handle:0x81820000". That is unless a hardcoded handle location was defined by each switch vendor and refer to as tpm-custom-key-1 for instance and used as a flag.

You could imagine TLS credential key/cert stored in the TPM owner hierarchy instead of file system and referenced using those handle locations that could be user defined then:
key = "handle:0x81820000"
cert = "handle:0x1C90000"

Another approach could be to have a switch defined hardcoded handle for such a use-case and refer to for instance "custom-key" in TLS profile for the system to know what to use.

Reply:
so we definitely want to abstract out such a specific address.
it should just be we have certs and or keys which can be used by other services on the device here is the "abstract" location of those materials (i.e. ssl_profile_id)

in the idevid / oidevid cases it does seem like that they don't quite fit into this model without some specific profiles which then should be what we setup so it is easy for users to use

Reply:
regarding the previous comment, note that the custom address scenario is not applicable to your use case, at least for now. If we do a generic enum, it can be extended later to support this use case.

@brianneville indeed our thinking is about a general modelling of tls profiles, not directly related to any interface (although this conversation was initiated due to a specific workflow).   
one thing that we need to consider is that tpm storage can be applicable not only to keys, but to certificates as well (in idevid/oidevid scenarios specifically). hence the enum should not be limited to the `key` field.

Reply:
>one thing that we need to consider is that tpm storage can be applicable not only to keys, but to certificates as well (in idevid/oidevid scenarios specifically). hence the enum should not be limited to the key field.

Maybe we could capture this in the Certificate message type like this:
```proto
message Certificate {
  // Type of certificate.
  CertificateType type = 1;
  // Certificate encoding type.
  CertificateEncoding encoding = 2;

  // Actual certificate.
  // The exact encoding depends upon the type of certificate.
  // for X509, this should be a PEM encoded Certificate.
  bytes certificate = 3;
  
  // Optional.
  // Not needed if the device knows the private key or does not have to know it.
  bytes private_key = 4;
  
  enum KeySource {
     KEYSOURCE_UNSPECIFIED = 0;
     KEYSOURCE_GENERATED = 1; // case where certz generated the key with a CSR
     KEYSOURCE_TRANSFER = 2; // case where key is transferred in this message ( private_key != nil )
     
     // keys corresponding to specific certs
     KEYSOURCE_OIDEVID = 3; 
     KEYSOURCE_IDEVID = 4; 
  }
  
  KeySource key_source = 5;

  enum CertSource {
     CERTSOURCE_UNSPECIFIED = 0;
     CERTSOURCE_TRANSFER = 1; // case where cert is transferred in this message ( certificate != nil )
     
     CERTOURCE_OIDEVID_TPM = 3; // cert is oidevid stored in TPM
     CERTOURCE_IDEVID_TPM = 4; // cert is idevid stored in TPM
  }
  
  CertSource cert_source = 6;
}
```

Reply:
i think this is where we are headed 
though might use a oneof rather than the enum and + certi/ key fields 


oneof i believe would make this simpler code wise

Reply:
sounds good to me!

Reply:
https://github.com/openconfig/gnsi/pull/172